src/devices/wifi/nm-device-iwd.c

```diff
@@ -1746,38 +1872,34 @@ nm_device_iwd_set_dbus_object (NMDeviceIwd *self, GDBusObject *object)
 }
 
 gboolean
-nm_device_iwd_agent_psk_query (NMDeviceIwd *self,
-                               GDBusMethodInvocation *invocation)
+nm_device_iwd_agent_query (NMDeviceIwd *self,
+                           GDBusMethodInvocation *invocation)
 {
 	NMDeviceIwdPrivate *priv = NM_DEVICE_IWD_GET_PRIVATE (self);
 	NMActRequest *req;
-	NMSettingWirelessSecurity *s_wireless_sec;
-	const gchar *psk;
+	const gchar *setting_name;
+	const gchar *setting_key;
+	gboolean replied;
 
 	req = nm_device_get_act_request (NM_DEVICE (self));
 	if (!req)
 		return FALSE;
 
-	s_wireless_sec = nm_connection_get_setting_wireless_security (nm_act_request_get_applied_connection (req));
-	if (!s_wireless_sec)
+	if (!try_reply_agent_request (self, nm_act_request_get_applied_connection (req),
+	                              invocation, &setting_name, &setting_key,
+	                              &replied))
 		return FALSE;
 
-	psk = nm_setting_wireless_security_get_psk (s_wireless_sec);
-	if (psk) {
-		_LOGD (LOGD_DEVICE | LOGD_WIFI,
-		       "Returning the PSK to the IWD Agent");
-
-		g_dbus_method_invocation_return_value (invocation,
-		                                       g_variant_new ("(s)", psk));
+	if (replied)
 		return TRUE;
-	}
 
 	nm_device_state_changed (NM_DEVICE (self), NM_DEVICE_STATE_NEED_AUTH,
 	                         NM_DEVICE_STATE_REASON_NO_SECRETS);
-	wifi_secrets_get_secrets (self,
-	                          NM_SETTING_WIRELESS_SECURITY_SETTING_NAME,
-	                          NM_SECRET_AGENT_GET_SECRETS_FLAG_ALLOW_INTERACTION
-	                            | NM_SECRET_AGENT_GET_SECRETS_FLAG_REQUEST_NEW);
+	wifi_secrets_get_one (self,
+	                      setting_name,
+	                      NM_SECRET_AGENT_GET_SECRETS_FLAG_ALLOW_INTERACTION
+	                        | NM_SECRET_AGENT_GET_SECRETS_FLAG_REQUEST_NEW,
+	                      setting_key);
 
 	priv->secrets_request = invocation;
```

Comment:
it seems a bit error-prone to piggyback the invocation on `priv`. Indeed, you only set `priv->secrets_request` when also schedule a new secret request (that is: `priv->wifi_secrets_id` is initialized), and cancelling this request will (should) cleanup the invocation. It's not immediately obvious, that `wifi_secrets_get_one()` does nothing but setting `priv->wifi_secrets_id`, but it's correct. All of this is almost correct, but confusing.

A small leak is that `wifi_secrets_cancel()` will call `nm_act_request_cancel_secrets()`, which calls `wifi_secrets_cb()`, which clears `priv->secrets_request = NULL;`, thus the line in `wifi_secrets_cancel()` will never send `g_dbus_method_invocation_return_error_literal()`. The simple solution for that, is to move
```
»···if (priv->secrets_request) {
»···»···g_dbus_method_invocation_return_error_literal (priv->secrets_request, NM_DEVICE_ERROR,
»···»···                                               NM_DEVICE_ERROR_INVALID_CONNECTION,
»···»···                                               "NM secrets request cancelled");
»···»···priv->secrets_request = NULL;
»···}
```
into `wifi_secrets_cb()`.

But while at it, could you not piggyback priv->secret_requests? Instead, pass it on as async data. As you then have more async-data, you could use `nm_utils_user_data_pack()` / `nm_utils_user_data_unpack()` [[1]](https://cgit.freedesktop.org/NetworkManager/NetworkManager/tree/src/nm-manager.c?id=f4bef9a12eba69e2cd94b2428939d2a523f4fce0#n4869). I prefer that, because then the invocation is always explicitly passed on and it's clearer to follow (instead of hooking it at `priv`).

Reply:
Oh yes, it was probably good before and I broke it with this path.  Now moved to use the suggested `nm_utils_user_data_pack()` / `nm_utils_user_data_unpack()` utilities in a separate preceding commit.
Also added one more unrelated fix as a new patch.